serve.go

```diff
@@ -24,26 +27,45 @@ func (s *Server) Run() error {
 // It also generates the OpenAPI spec and outputs it to a file, the UI, and a handler (if enabled).
 func (s *Server) RunTLS(certFile, keyFile string) error {
 	s.isTLS = true
-
-	s.setup()
-	return s.Server.ListenAndServeTLS(certFile, keyFile)
+	if err := s.setup(); err != nil {
+		return err
+	}
+	return s.Server.ServeTLS(s.listener, certFile, keyFile)
 }
 
-func (s *Server) setup() {
+func (s *Server) setup() error {
+	if err := s.setupDefaultListener(); err != nil {
+		return err
+	}
 	go s.OutputOpenAPISpec()
 	s.printStartupMessage()
 
 	s.Server.Handler = s.Mux
 	if s.corsMiddleware != nil {
 		s.Server.Handler = s.corsMiddleware(s.Server.Handler)
 	}
+
+	return nil
+}
+
+func (s *Server) setupDefaultListener() error {
+	if s.listener != nil {
+		WithAddr(s.listener.Addr().String())(s)
+		return nil
+	}
+	listener, err := net.Listen("tcp", s.Addr)
+	if err != nil {
+		return err
+	}
+	WithListener(listener)(s)
```

Comment:
Why?
WithListener sets s.listener, but setting a s.listener will also trigger WithListener that will re-set s.listener with the same value? The flow seems a bit weird to me but I might miss something, please tell me :)

Reply:
This is for the case where a user has provided a listener. It would probably be more clear if it was

s.listener = listener than using the With function.